src/Deinterleave.cpp

```diff
@@ -176,224 +178,637 @@ Stmt collect_strided_stores(const Stmt &stmt, const std::string &name, int strid
     return collect.mutate(stmt);
 }
 
+// Essentially, broadcast the mask, i.e. take each mask element, and repeat it
+// \p scale times, consequtively. No reordering of mask elements happens!
+std::vector<char> upscale_mask(const std::vector<char> &src_mask, int scale) {
+    internal_assert(scale > 1) << "Expected to at least double the mask.";
+    std::vector<char> mask;
+    mask.reserve(scale * src_mask.size());
+    for (char src_mask_elt : src_mask) {
+        std::fill_n(std::back_inserter(mask), /*count=*/scale,
+                    /*value=*/src_mask_elt);
+    }
+    return mask;
+}
+
+// Compute the elements of the specified ramp/linear space.
+std::vector<int> ramp_to_indices(int starting_lane, int lane_stride,
+                                 int new_lanes) {
+    std::vector<int> indices;
+    indices.reserve(new_lanes);
+    for (int i = 0; i < new_lanes; i++) {
+        indices.emplace_back(starting_lane + lane_stride * i);
+    }
+    return indices;
+}
+
+// What are the (in-order) demanded indices?
+std::vector<int> mask_to_indices(const std::vector<char> &mask) {
```

Comment:
I'd a little bit confused about why this PR uses lists of demanded lanes but also masks. Why are both representations necessary?

Reply:
Neither is optimal for the problem at hand.
For example, it is `O(1)` to compute the number of demanded lanes given the list (vector) of demanded lanes,
but if you have a mask of demanded lanes, it's `O(N)`.
Likewise, given a mask, it is trivial to upscale/downscale it (repeat/or-reduce bits),
but given indices it becomes somewhat less straight-forward.
Etc, etc.

We //could// settle for one, but code like this is exceptionally easy to be suddenly broken,
which is why i've intentionally written the diff as it is now, very verbose.

Reply:
I think upscaling indices is:
```
vector<int> upscale(const vector<int> &v, int scale) {
  vector<int> result(v.size() * scale);
  for (size_t i = 0; i < v.size(); i++) {
    for (int s = 0; s < scale; s++) {
      result[i*scale + s] = v[i] * scale + s;
    }
  }
  return result;
}
```
Which doesn't seem that much more complex than upscaling a mask. Can we switch to just indices?

Reply:
Okay, i've rewritten `upscale_indices()`.

Though, that doesn't help much, because `IndexXFormMapping::get()` still does that roundtrip,
because the alternative is using a set/sorting a vector, and i can't possibly imagine that being better...

Reply:
If that all it's for now, can it be made a private method in IndexXFormMapping called, e.g. sort_and_dedup_indices?

Reply:
I was proposing reducing the scope in which it's defined, and renaming it to make clear what its sole usage is, so that people reading the file understand more clearly what it's for. But if you're saying there are still other uses, then I guess don't do that.

Another minor comment on naming: I think "affine" is more common than "1st_order_polynomial" in compilers, and it's a little terser.

Reply:
(I ended up not needing it in another place, all hail #6908!)

My main concern is that said function has nothing specific to do with it's (only) user.
By that logic, we should also drop other single-use functions.
But to me that only hurts - they are there because they have obvious semantics,
and are separateable from users. Abstractions all the way, it's the only way to try to not sink.